hathor/p2p/sync_mempool.py

```diff
@@ -0,0 +1,173 @@
+    from hathor.p2p.node_sync_v2 import NodeBlockSync
+
+logger = get_logger()
+
+
+class SyncMempoolManager:
+    """Manage the sync-v2 mempool with one peer.
+    """
+    def __init__(self, sync_manager: 'NodeBlockSync'):
+        """Initialize the sync-v2 mempool manager."""
+        # Shortcuts.
+        self.sync_manager = sync_manager
+        self.manager = self.sync_manager.manager
+        self.tx_storage = self.manager.tx_storage
+        self.reactor = self.sync_manager.reactor
+
+        # Set of tips we know but couldn't add to the DAG yet.
+        self.missing_tips: Set[bytes] = set()
+
+        # Stack used by the DFS in the dependencies.
+        # We use a deque for performance reasons.
+        self.deque: Deque[BaseTransaction] = deque()
+
+        # Maximum number of items in the DFS.
+        self.MAX_STACK_LENGTH: int = 1000
+
+        # Internal variables.
+        # This is used to keep track of the latest requested tx.
+        self._download_tx_deferred: Optional[Deferred] = None
+        self._download_tx_hash: Optional[bytes] = None
+
+    def start(self):
+        pass
+
+    @inlineCallbacks
+    def get_tx(self, tx_id: bytes) -> Generator[BaseTransaction, BaseTransaction, BaseTransaction]:
+        """Async internal method to get a transaction from the db or to download it."""
+        try:
+            tx = self.tx_storage.get_transaction(tx_id)
+        except TransactionDoesNotExist:
+            tx = yield self._download_tx(tx_id)
+        return tx
+
+    @inlineCallbacks
+    def _download_tx(self, tx_id: bytes) -> Generator[BaseTransaction, BaseTransaction, BaseTransaction]:
+        """Async internal method to download a transaction."""
+        tx = yield self._get_data(tx_id)
+        # TODO Verify tx.
+        # tx.verify()
+        return tx
+
+    def _get_data(self, tx_id: bytes) -> Deferred:
+        """Internal method to request the tx."""
+        assert self._download_tx_deferred is None
+        assert self._download_tx_hash is None
+        self._download_tx_deferred = Deferred()
+        self._download_tx_hash = tx_id
+        self.sync_manager.send_get_data(tx_id, origin='mempool')
+        return self._download_tx_deferred
+
+    def on_new_mempool_tx(self, tx: BaseTransaction) -> None:
+        """Called when a requested tx is received."""
+        assert self._download_tx_deferred is not None
+        if tx.hash != self._download_tx_hash:
+            # Peer sent the wrong transaction?!
+            # Maybe we should ban the peer.
+            raise Exception('This should never happen.')
+        deferred = self._download_tx_deferred
+        self._download_tx_deferred = None
+        self._download_tx_hash = None
+        deferred.callback(tx)
+
+    def run(self) -> None:
+        """Run a loop of the sync-v2 mempool and schedules the next call.
+        It can safely be called multiple times.
+        """
+        self.run_once()
+        self.reactor.callLater(1000, self.run)
```

Comment:
I guess it should be called every `1` second (instead of `1000` seconds). And it would be great to move it to an attribute in the class.

Reply:
This was refactored.

Reply:
Ok. Thanks!